[groups persons] API for returning groups based on trend results

EDsCODE opened the pull request:
## Changes

*Please describe.*  
- breaking ground on adding ability to view groups when clicking on datapoint
- adds functionality for building a group query based on unique groups trend math
- moves test and adds persons querying to preexisting aggregate test
*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*
- moved/added test

EDsCODE committed:
working for unique_groups math

EDsCODE committed:
fix types

EDsCODE committed:
add null check

EDsCODE committed:
update snapshots

EDsCODE committed:
update payload

EDsCODE committed:
update snapshots

EDsCODE commented:
@macobo feel free to give it an initial pass to make sure it doesnt collide/violate any patterns you had in mind!

macobo commented on ee/clickhouse/queries/trends/person.py:
```diff
@@ -76,8 +82,39 @@ def get_query(self) -> Tuple[str, Dict]:
+        else:
+            return self.get_person_query()
+
     def get_people(self) -> ReturnDict:
```
Can we add a TypedDict return type here?

macobo reviewed (changes requested):
Left code-wise comments.

I am not a fan of the solution on a high level either though: It feels we've doubled the amount of queries we need to write and manage.

IMO the problems to solve are:
1. What to name groups + person query together? From this follows class/instance naming, endpoints, etc.
2. Can we unify the pattern for querying? E.g. something like this in a base class everything can use
```python
def get_query():
    event_query, event_params = ...
    if self.aggregating_by_groups: 
        query, params = self.groups_query()
        return query.format(...), { *event_params, *params }
    else:
        # same thing
```
    And have the `groups_query`/`person_query` be as standard as possible across all insights?
3. Can we unify the pattern for serializing? E.g. the return format is similar except for a few values and one literal which decides what we'll render in the FE. 
4. We should put a lot of effort into proper typing of return values.

I'm also confused at the lack of "links" along with datapoints in the code, but I might be missing context.

macobo committed:
Add group model

```sql
BEGIN;
--
-- Create model Group
--
CREATE TABLE "posthog_group" ("id" serial NOT NULL PRIMARY KEY, "group_key" varchar(400) NOT NULL, "group_type_index" integer NOT NULL, "group_properties" jsonb NOT NULL, "created_at" timestamp with time zone NOT NULL, "properties_last_updated_at" jsonb NOT NULL, "properties_last_operation" jsonb NOT NULL, "version" bigint NOT NULL, "team_id" integer NOT NULL);
--
-- Create constraint unique team_id/group_key/group_type_index combo on model group
--
ALTER TABLE "posthog_group" ADD CONSTRAINT "unique team_id/group_key/group_type_index combo" UNIQUE ("team_id", "group_key", "group_type_index");
ALTER TABLE "posthog_group" ADD CONSTRAINT "posthog_group_team_id_b3aed896_fk_posthog_team_id" FOREIGN KEY ("team_id") REFERENCES "posthog_team" ("id") DEFERRABLE INITIALLY DEFERRED;
CREATE INDEX "posthog_group_team_id_b3aed896" ON "posthog_group" ("team_id");
COMMIT;
```

macobo committed:
Remove a dead import

macobo committed:
Improve typing for groups

macobo committed:
Make groups updating more generic, avoid mutation

This simplifies using the same logic for groups

Note there's a behavioral change: We don't produce a new kafka message
if nothing has been updated anymore.

macobo committed:
Rename a function

macobo committed:
WIP: Handle group property updates

... by storing them in postgres

Uses identical pattern to person property updates, except we handle
first-seen case within updates as well.

macobo committed:
Get rid of boolean option

macobo committed:
WIP continued

EDsCODE commented on ee/clickhouse/views/test/test_clickhouse_trends.py:
```diff
@@ -332,3 +320,67 @@ def test_insight_trends_compare(self):
         )
+
+
+class ClickhouseTestTrendsGroups(ClickhouseTestMixin, LicensedTestMixin, APIBaseTest):
```
Moved test here because i changed it to an API test. Also, added a different class just so it's clearer and we can move it to a different file just for organizations sake

EDsCODE commented on ee/clickhouse/views/test/test_clickhouse_trends.py:
```diff
@@ -332,3 +320,67 @@ def test_insight_trends_compare(self):
+        )
+
+    @snapshot_clickhouse_queries
+    def test_aggregating_by_group(self):
```
Do you mean testing the persons alongside the data or the multiple group type index filtering? Part of the improvements we were making last week was to connect the person querying with the data querying since the numbers should always match and that's relevant in testing them together

macobo replied:
I mean it's doing two different `TrendsRequest`s below and on top of that has person logic following that.

Trying to do too many things in a test often causes it to become very hard to understand. Then when it breaks it gets deleted or "hacked around" rather than being properly handled.

EDsCODE replied:
removed second part

EDsCODE commented on ee/clickhouse/sql/person.py:
```diff
@@ -317,3 +317,19 @@
+        group_created_at_{group_type_index},
+        group_properties_{group_type_index}
+    FROM ({events_query})
+    GROUP BY 
```
I noticed this too and it exists in the trends query which was really confusing. I opted to just stick to almost exactly how the trend query is built for consistency. But now I can refactor them together

EDsCODE commented on ee/clickhouse/sql/person.py:
```diff
@@ -317,3 +317,19 @@
+        group_created_at_{group_type_index},
+        group_properties_{group_type_index}
+ LIMIT 200
+OFFSET 0
```
unfinished

EDsCODE committed:
use constructor

EDsCODE commented on ee/clickhouse/queries/groups_join_query.py:
```diff
@@ -1,9 +1,19 @@
+CREATED_AT = ColumnName("created_at")
+GROUP_PROPERTIES = ColumnName("group_properties")
+ACCEPTED_EXTRA_FIELDS = [CREATED_AT]
+GROUP_ALIASES = {
```
I added this type of query building so that [this select](https://github.com/PostHog/posthog/pull/7144/files#diff-68ee5fd398e73fa594adb5050743b2d33a3e9e407ca1444370432c49a18e6677R52) in the above query will be flexible and able to account for different fields of different group types. 

Also, if there's a query that's filtering on multiple group types, there will be multiple group joins so `Dict[GroupTypeIndex, ColumnName]` type allows you to specify which group join to add the fields in and for the query above to select the right one. 

Open to simplifying as you suggest but I think it would make the select in the outer query have more assumptions with regards to which group type index to select on

macobo replied:
> I think it would make the select in the outer query have more assumptions with regards to which group type index to select on

Let's take a look at the usecase:

```python
            join_queries.append(
                f"""
                INNER JOIN (
                    SELECT
                        group_key,
                        argMax(group_properties, _timestamp) AS group_properties_{group_type_index}
                        {(
                            " ".join(
                                f", argMax({field}, _timestamp) AS {GROUP_ALIASES[field]}{group_type_index}"
                                for field in columns_to_add
                            )
                        )}
                    FROM groups
                    WHERE team_id = %(team_id)s AND group_type_index = %({var})s
                    GROUP BY group_key
...
```

The problem I see it's this constant makes it becomes almost impossible to understand what this will expand to. One would need to:
- Jump to see what `_extra_fields` are passed
- Jump to see what `ACCEPTED_EXTRA_FIELDS` are, do the set logic
- Jump to constant to see what the mapping is
- Jump to query and see we're implicitly appending {group_type_index} at the end of it

We'd be jumping between so many different files for confusing reasons, making this hard to grok.

Let's discuss a potential solution. Perhaps we should be doing 

```
SELECT columns
FROM groups
WHERE id IN ({events_query})
```

instead and not bring this complexity within event queries and into all classes in the first place?

EDsCODE replied:
Resolved on call: We decided to always just retrieve IDs of groups or persons and retrieve them using postgres data. This way it's consistent across all insight implementations and doesn't require as much query reshaping to retrieve the right fields

EDsCODE committed:
adjust queries

EDsCODE committed:
introduce base class

EDsCODE committed:
consolidate querying

EDsCODE committed:
shared serializer and typed

EDsCODE committed:
merge master

EDsCODE committed:
sort imports

EDsCODE committed:
snapshots

EDsCODE committed:
typing

EDsCODE committed:
change name

EDsCODE commented:
Doesn't handle breakdown filtering yet but otherwise good for a review

macobo committed:
fetchGroup() and upsertGroup()

macobo committed:
Test more edge cases

macobo committed:
Add tests for upsertGroup() in properties-updater

macobo committed:
Rename to PropertyUpdateOperation

macobo committed:
Followup

macobo committed:
Solve typing issues

macobo commented on ee/clickhouse/client.py:
```diff
@@ -145,8 +145,12 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
+                    prepared_sql,
+                    params=prepared_args,
+                    settings=settings,
+                    with_column_types=True if as_dict else with_column_types,
```
`as_dict or with_column_types`

macobo commented on ee/clickhouse/client.py:
```diff
@@ -166,6 +170,15 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
                     save_query(prepared_sql, execution_time)
         return result
 
+    def _convert_to_dict(result) -> List[Dict[str, Any]]:
```
This is pretty incomprehensible without better naming. Perhaps extract `header_row = result[1]` or smth?

macobo commented on ee/clickhouse/client.py:
```diff
@@ -135,7 +135,7 @@ def cache_sync_execute(query, args=None, redis_client=None, ttl=CACHE_TTL, setti
             return result
 
-    def sync_execute(query, args=None, settings=None, with_column_types=False):
+    def sync_execute(query, args=None, settings=None, with_column_types=False, as_dict=False):
```
Hmm. This muddies the return type of this base function a lot - it doesn't feel like we want even more options here.

Let's discuss this through during pairing.

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,70 @@
+        created_at = data.get("created_at", None)
+        properties = json.loads(data.get("properties", "{}"))
+        is_identified: Optional[bool] = data.get("is_identified", None)
+        alias: str = properties.get("email", None) or properties.get("name", None)
```
This only makes sense for persons, not groups. Groups have `group_key`

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,70 @@
+        self.entity = entity
+        self.filter = filter
+
+        if self.entity and self.entity.math == "unique_group":
```
Comment got buried: I don't think this should be handled this way and instead either inlined or extracted into a property.

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,70 @@
+from posthog.models import Entity, Filter, Team
+
+
+class ActorResponse(TypedDict):
```
This really isn't a `Response`, but only a subset of one.

Also there's no real way for the frontend to know given this object if it's dealing with a group or person, which will cause pain.

Rather than a single dict I think this would be much clearer as an union of two:

```
class SerializedPerson(TypedDict):
    type: Literal['person']
    ...

class SerializedGroup(TypedDict):
    type: Literal['group']
   ...

SerializedActor = Union[SerializedPerson, SerializedGroup]
```

macobo commented on ee/clickhouse/queries/trends/person.py:
```diff
@@ -36,16 +38,21 @@ def _handle_date_interval(filter: Filter) -> Filter:
-        if self.filter.display != TRENDS_CUMULATIVE and not self.filter.display in TRENDS_DISPLAY_BY_VALUE:
-            self.filter = _handle_date_interval(self.filter)
+    def __init__(self, team: Team, entity: Optional[Entity], filter: Filter):
+        new_filter = filter
```
Nit: This temporary var isn't really needed

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,70 @@
+    distinct_ids: List[str]
+
+
+class ActorBaseQuery:
```
:+1: for this!

macobo commented on posthog/api/utils.py:
```diff
@@ -19,16 +20,37 @@ class PaginationMode(Enum):
 
 
 def get_target_entity(request: request.Request) -> Entity:
```
Q: What caused this change? I didn't see anything else relating to this changing.

EDsCODE replied:
Retrieving groups from trends test was failing because this helper function was building the entity using solely ENTITY_TYPE and ENTITY_MATH but didn't have the group type index

EDsCODE committed:
Merge branch 'groups-update-support' into trends-groups-person-query-backend

EDsCODE committed:
changed implementation to use pg

EDsCODE committed:
unusd

EDsCODE committed:
update type

EDsCODE committed:
update snapshots

EDsCODE commented on ee/clickhouse/models/group.py:
```diff
@@ -2,12 +2,27 @@
+
+    @receiver(post_save, sender=Group)
+    def person_created(sender, instance: Group, created, **kwargs):
+        create_group(
```
mirrors the person implementation so that when groups are created from the django orm (especially during tests), the clickhouse and postgres data will match

macobo commented on ee/clickhouse/models/group.py:
```diff
@@ -2,12 +2,27 @@
+if is_clickhouse_enabled():
+
+    @receiver(post_save, sender=Group)
+    def person_created(sender, instance: Group, created, **kwargs):
```
Nit: Name of the function is wrong.

We want to avoid this pattern if we can - this method won't be called in app code ever and if it is it's guaranteed to be wrong (e.g. properties_last_updated and properties_last_operation are missing), which will cause further bugs down the line.

I assume this is there for tests? If so, worth updating the `create_group` helper instead and perhaps update the name/file location to indicate this is test-specific?

EDsCODE replied:
Hmm I added the test flag to it. Not quite sure where to move this logic though where it will still be obvious

macobo commented on ee/clickhouse/queries/column_optimizer.py:
```diff
@@ -104,7 +105,10 @@ def properties_used_in_filter(self) -> Counter[PropertyIdentifier]:
             # See ee/clickhouse/queries/trends/util.py#process_math
             if entity.math == "unique_group":
-                counter[(f"$group_{entity.math_group_type_index}", "event", None)] += 1
+                if self.is_actor_query:
```
Not sure this if-else is correct.

It's saying if it's an actor query, property `$group_{N}` for group type math_group_type_index is used, which does not seem correct.

macobo commented on ee/clickhouse/queries/column_optimizer.py:
```diff
@@ -19,9 +19,10 @@ class ColumnOptimizer:
     """
 
-    def __init__(self, filter: Union[Filter, PathFilter, RetentionFilter], team_id: int):
+    def __init__(self, filter: Union[Filter, PathFilter, RetentionFilter], team_id: int, is_actor_query: bool = False):
```
Q: Why is this needed?

In general we want to avoid coupling column_optimizer with other modules even tighter. Since everything already depends on it, we end up with a two-way dependency which is just hell to debug.

EDsCODE replied:
Didn't like the coupling either but seemed like the most streamlined way to add in the flag that would tell the querybuilder to add the group join in

EDsCODE replied:
nvm. was looking at this the wrong way. after we decided to use postgres data, very little changes were needed

macobo commented on ee/clickhouse/queries/event_query.py:
```diff
@@ -103,11 +104,6 @@ def _determine_should_join_persons(self) -> None:
             self._should_join_persons = True
             return
 
-        if self._filter.breakdown_type == "person":
```
Q: Why remove this? Was this handled by column_optimizer already?

EDsCODE replied:
no idea i think it got deleted when i merged in your branch because of previously edited deletions

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,114 @@
+                type="group", group_key=group.group_key, created_at=group.created_at, properties=group.group_properties
+            )
+            for group in data
+            if isinstance(group, Group)
```
Can it ever not be Group here? I think not right?

If the goal is type-casting then I'd change the signature to

```
    def _serialize_groups(self, data: QuerySet[Group]) -> List[SerializedActor]:
```

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,114 @@
+
+    def _get_people(self, results) -> Tuple[QuerySet[Actor], List[SerializedActor]]:
+        """ Get people from raw SQL results in data model and dict formats """
+        persons: QuerySet[Actor] = Person.objects.filter(team_id=self.team.pk, uuid__in=[val[0] for val in results])
```
See below comment on the _serialize_people signature. If you annotate this as `persons: QuerySet[Person]` everything should work well.

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,114 @@
+        if self.entity and self.entity.math == "unique_group":
+            self.aggregating_by_groups = True
+
+    def groups_query(self) -> Tuple[str, Dict]:
```
Q: Why do subclasses need to implement both? Existing queries are already using `aggregation_target` as the column to return from event_query, which would be the group key in one case and person_id in the other.

macobo replied:
That said, looking at the only implementation these are really similar. I'll need to read through the code with a fresh head tomorrow, I _think_ there might be ways to simplify this further.

EDsCODE replied:
Yup you're right. I combined the functions

EDsCODE committed:
rename and remove inlining

EDsCODE committed:
restore bad merge code

EDsCODE committed:
adjust types

EDsCODE committed:
add flag

EDsCODE committed:
remove var

EDsCODE committed:
misnamed

EDsCODE committed:
change to uuid

EDsCODE committed:
make sure to use string when passing result

EDsCODE committed:
remove from columnoptimizer logic and have group join logic implemented by event query classes per insight

EDsCODE committed:
remove unnecessary logic

EDsCODE committed:
typing

EDsCODE commented on ee/clickhouse/queries/trends/person.py:
```diff
@@ -74,19 +80,20 @@ def get_query(self) -> Tuple[str, Dict]:
-            extra_person_fields=["created_at", "person_props", "is_identified"],
+            should_join_distinct_ids=False if self.is_aggregating_by_groups else True,
+            should_join_persons=False if self.is_aggregating_by_groups else True,
+            extra_fields=[] if self.is_aggregating_by_groups else ["distinct_id", "team_id"],
```
Realized this could just be a single query after we agreed we can just pull data from postgres with the ids. There's very few extra params needed when it's a group query besides these 3. No need to push an extra parameters down to columnoptimizer or the eventquery builder

EDsCODE reviewed (commented):
Ended up needing very few changes. Most of the diffs are from tests and snapshot changes

macobo commented on ee/clickhouse/queries/actor_base_query.py:
```diff
@@ -0,0 +1,106 @@
+
+    @cached_property
+    def is_aggregating_by_groups(self) -> bool:
+        if self.entity and self.entity.math == "unique_group":
```
nit: `return self.entity is not None and self.entity.math == "unique_group"`

EDsCODE replied:
missed this but added it in the funnel PR

macobo commented on ee/clickhouse/queries/groups_join_query.py:
```diff
@@ -4,6 +4,7 @@
 from posthog.models import Filter
 from posthog.models.filters.path_filter import PathFilter
 from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.property import GroupTypeIndex
```
Nit: Dead import after changes.

macobo commented on ee/clickhouse/queries/event_query.py:
```diff
@@ -1,5 +1,5 @@
 from abc import ABCMeta, abstractmethod
-from typing import Any, Dict, List, Tuple, Union
+from typing import Any, Dict, List, Optional, Set, Tuple, Union
```
Nit: There's some new dead imports here

macobo commented on ee/clickhouse/queries/trends/person.py:
```diff
@@ -74,19 +80,20 @@ def get_query(self) -> Tuple[str, Dict]:
-            extra_fields=["distinct_id", "team_id"],
-            extra_person_fields=["created_at", "person_props", "is_identified"],
+            should_join_distinct_ids=False if self.is_aggregating_by_groups else True,
+            should_join_persons=False if self.is_aggregating_by_groups else True,
```
Nit: `should_join_persons=not self.is_aggregating_by_groups` is imo cleaner :)

macobo commented on ee/clickhouse/queries/trends/person.py:
```diff
@@ -36,16 +39,19 @@ def _handle_date_interval(filter: Filter) -> Filter:
-        self.team = team
-        self.entity = entity
-        self.filter = filter
+class TrendsPersonQuery(ActorBaseQuery):
```
Note: We should rename this + endpoints in future PRs as well.

macobo commented on ee/clickhouse/sql/person.py:
```diff
@@ -299,21 +299,11 @@
-    arrayReduce('groupUniqArray', groupArray(distinct_id)) AS distinct_ids
+GET_ACTORS_FROM_EVENT_QUERY = """
+SELECT 
+    {id_field} AS id
```
Nit: let's name this `AS actor_id` - this way we can be consistent in other queries. id is IMO overloaded.

macobo reviewed (approved):
:heart_eyes: This took some iteration but really happy where this ended up - I think this is quite a sustainable pattern for the future.

macobo commented on ee/clickhouse/models/group.py:
```diff
@@ -2,12 +2,28 @@
+if is_clickhouse_enabled() and settings.TEST:
+
+    @receiver(post_save, sender=Group)
+    def group_created(sender, instance: Group, created, **kwargs):
```
I'd move this inside the `create_group` function. It's only ever used in tests.

EDsCODE committed:
remove dead imports

EDsCODE committed:
remove verbosity

EDsCODE committed:
update snapshots

EDsCODE committed:
merge master

EDsCODE committed:
typos

EDsCODE committed:
remove signals

EDsCODE committed:
remove plugin excess
